### PR TITLE
LPD-18760 - Retrieve FDS views information from DSM

### PR DESCRIPTION
**Pull request by juanjofgliferay**
## Story/Task
[LPD-6622 - Allow users to change visualization mode](https://liferay.atlassian.net/browse/LPD-6622) / [LPD-18760](https://liferay.atlassian.net/browse/LPD-18760)

## Scope
Retrieve the information that an admin user provides in the DSM when configuring a data set and display available visualization options in the Frontend Dataset component
- get mapped fields for each visualization mode
- filter out visualization modes without mapped fields
- default to table visualization mode if no one has mapped fields

## Concerns
The `Cards` component needs some love :cupid: 
Solved a couple of quick issues in this PR:
- provide a spritemap and a symbol to display the list of available actions
- hide the 3 dots menu if no action is defined

## UI
[Screencast from 2024-02-29 17-19-43.webm](https://github.com/liferay-frontend/liferay-portal/assets/132653342/a36fcc4a-5578-45df-8394-b8b189d0e4a4)

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 1]
```diff
@@ -260,6 +260,20 @@ private String _buildFragmentHTML(
...
+			fdsCardObjectEntries = _getRelatedObjectEntries(
+				fdsViewObjectDefinition, fdsViewObjectEntry,
+				"fdsViewFDSCardsSectionRelationship");
+
+			fdsListObjectEntries = _getRelatedObjectEntries(
+				fdsViewObjectDefinition, fdsViewObjectEntry,
+				"fdsViewFDSListSectionRelationship");
+		}
+
 		Set<ObjectEntry> fdsFieldObjectEntries = _getFDSFieldObjectEntries(
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:277]: Maybe it is time to rename this `fdsFieldObjectEntries`  to `fdsTableObjectEntries`

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 2]
```diff
@@ -433,6 +458,89 @@ private Set<ObjectEntry> _getFDSFieldObjectEntries(
...
+			Collection<ObjectEntry> fdsListViewObjectEntries,
+			HttpServletRequest httpServletRequest)
+		throws Exception {
+
+		return JSONUtil.put(
+			"contentRenderer", "list"
+		).put(
+			"default", false
+		).put(
+			"label", _language.get(httpServletRequest, "list")
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:471]: Need to translate label here, as it didn't work in the frontend side :shrug:

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js [hunk 3]
```diff
@@ -29,7 +30,11 @@ const Cards = ({items, schema}) => {
 							className="col-md-3"
 							key={item[selectedItemsKey] || index}
 						>
-							<Card item={item} schema={schema} />
+							<Card
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js:33]: Just a quick fix to display item actions icons in the card component. This is something that we could provide up in the context and maybe sometime make it customizable.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java (hunk 1 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:277]: Don't tell anybody, but I'm planning to sneak this in as a part of the story that removes `FDSViews` :shushing_face:   I'm thinking `FDSTableColumn`, or even `DSMTableColumn`.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js (hunk 3 above)
Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js:33]: Feel free to go nuts with `cards` view implementation. I don't think it's used anywhere in DXP, apart from the sample module. I suspect there are a lot of small issues like this, missing API implementations etc.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 4]
```diff
@@ -260,6 +260,20 @@ private String _buildFragmentHTML(
 			fragmentEntryLink.getCompanyId(), fdsEntryObjectEntryERC,
 			fdsEntryObjectDefinition);
 
+		Collection<ObjectEntry> fdsCardObjectEntries = null;
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:263]: It should be `fdsCardsSection*` and `fdsListSection*` throughout this PR, to match naming in persistence.

File: modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java [hunk 5]
```diff
@@ -433,6 +458,89 @@ private Set<ObjectEntry> _getFDSFieldObjectEntries(
...
+				viewsJSONArray.put(
+					_getFDSListViewJSONObject(
+						fdsListObjectEntries, httpServletRequest));
+			}
+
+			if (!fdsFieldObjectEntries.isEmpty()) {
+				viewsJSONArray.put(
+					_getFDSTableViewJSONObject(
+						companyId, fdsFieldObjectEntries, httpServletRequest));
+			}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java:529]: This is fine for now. We can tweak defaults later on.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js [hunk 6]
```diff
@@ -29,7 +30,11 @@ const Cards = ({items, schema}) => {
...
+								item={item}
+								schema={schema}
+								spritemap={SPRITEMAP}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js:36]: `spritemap` should not be needed, we provide it in context for Clay components. Not sure why this is not working in this specific component.

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay few comments inline

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js (hunk 6 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js:36]: Should I remove it?

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js [hunk 7]
```diff
@@ -74,6 +76,7 @@ const Card = ({item, schema}) => {
 						});
 					}
 				},
+				symbolLeft: action.icon,
```
Comment (juanjofgliferay) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js:79]: @markocikos Now it works with the context spritemap.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/constants.ts [hunk 8]
```diff
@@ -5,3 +5,6 @@
 
 export const FDS_NESTED_FIELD_NAME_DELIMITER: string = '.';
 export const FDS_NESTED_FIELD_NAME_PARENT_SUFFIX: string = `${FDS_NESTED_FIELD_NAME_DELIMITER}*`;
+
+const IMAGES_PATH = Liferay.ThemeDisplay.getPathThemeImages();
+export const SPRITEMAP = IMAGES_PATH + '/clay/icons.svg';
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/constants.ts:10]: Do we still need these?

**Review by markocikos — COMMENTED** [review]
@juanjofgliferay Apart from probably minor cleanup, LGTM!

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/constants.ts (hunk 8 above)
Comment (juanjofgliferay) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/constants.ts:10]: Ouch, I don't think so

**Review by markocikos — APPROVED** [review]

**Comment by juanjofgliferay** [comment]
ci:forward

**Comment by juanjofgliferay** [comment]
ci:report:2546100650

**Comment by juanjofgliferay** [comment]
Based on [TestRay report](https://github.com/liferay-frontend/liferay-portal/pull/4030#issuecomment-1976498320), failures are not related

**Comment by juanjofgliferay** [comment]
ci:forward:force

**Pull request closed without merging**